File: engine/src/warden/adapters/docker_adapter.py
"""Adapter docker — shell out pra `docker compose` (não docker-py), um stack por projeto."""

import json
import subprocess

from warden.adapters.base import Adapter, ProcessStatus
from warden.config import ProjectConfig
from warden.vitals import VitalsSampler


class DockerAdapter(Adapter):
    """`compose_services` vazio = compose file é só desse projeto, comando afeta o
    arquivo inteiro (comportamento antigo). Preenchido = compose file compartilhado
    por vários projetos (stack único por workspace) — todo comando fica restrito
    aos serviços listados, pra não subir/derrubar/misturar log dos vizinhos."""

    def __init__(self, config: ProjectConfig):
        self.config = config
        self._compose_file = config.compose_file or "docker-compose.yml"
        self._services = config.compose_services
        self._vitals = VitalsSampler()

    def _compose(self, *args: str) -> subprocess.CompletedProcess:
        return subprocess.run(
            ["docker", "compose", "-f", self._compose_file, *args],
            cwd=self.config.path,
            capture_output=True,
            text=True,
        )

    def start(self) -> None:
        self._compose("up", "-d", *self._services)

    def stop(self) -> None:
        self._compose("stop", *self._services)

    def status(self) -> ProcessStatus:
        result = self._compose("ps", "--format", "json")
        containers = _parse_ps_json(result.stdout)
        if self._services:
            containers = [c for c in containers if c.get("Service") in self._services]
        running = [c for c in containers if c.get("State") == "running"]
        if not running:
            return ProcessStatus(running=False)
        pid = self._container_pid(running[0]["ID"])
        ports = sorted(
            {
                publisher["PublishedPort"]
                for container in running
                for publisher in container.get("Publishers") or []
                if publisher.get("PublishedPort")
            }
        )
        cpu_percent, memory_mb = None, None
        if pid:
            vitals = self._vitals.sample(pid)
            if vitals:
                cpu_percent, memory_mb = vitals
        return ProcessStatus(
            running=True, pid=pid, ports=ports, cpu_percent=cpu_percent, memory_mb=memory_mb
        )

    def _container_pid(self, container_id: str) -> int | None:
        result = subprocess.run(
            ["docker", "inspect", "-f", "{{.State.Pid}}", container_id],
            capture_output=True,
            text=True,
        )
        try:
            pid = int(result.stdout.strip())
        except ValueError:
            return None
        return pid or None

    def logs(self, tail: int = 100, service: str | None = None) -> list[str]:
        args = ["logs", "--no-color", "--tail", str(tail)]
        if service:
            args.append(service)
        elif self._services:
            args.extend(self._services)
        result = self._compose(*args)
        output = result.stdout + result.stderr
        return [line for line in output.splitlines() if line.strip()]

    def services(self) -> list[str]:
        if self._services:
            return self._services
        result = self._compose("config", "--services")
        return [line for line in result.stdout.splitlines() if line.strip()]


def _parse_ps_json(output: str) -> list[dict]:
    output = output.strip()
    if not output:
        return []
    try:
        data = json.loads(output)
        return data if isinstance(data, list) else [data]
    except json.JSONDecodeError:
        return [json.loads(line) for line in output.splitlines() if line.strip()]

File: ator/Warden.Domain/Adapters/DockerAdapter.cs
using System.Diagnostics;
using System.Text.Json;
using System.Text.Json.Serialization;
using Warden.Domain.Config;

namespace Warden.Domain.Adapters;

/// <summary>
/// Adapter docker — shell out pra `docker compose` (não a Docker Engine API), um stack por projeto.
///
/// `ComposeServices` vazio = compose file é só desse projeto, comando afeta o arquivo inteiro
/// (comportamento antigo). Preenchido = compose file compartilhado por vários projetos (stack único
/// por workspace) — todo comando fica restrito aos serviços listados, pra não subir/derrubar/misturar
/// log dos vizinhos.
/// </summary>
public sealed class DockerAdapter : IAdapter
{
    private static readonly JsonSerializerOptions JsonOptions = new() { PropertyNameCaseInsensitive = true };

    private readonly ProjectConfig _config;
    private readonly string _composeFile;
    private readonly IReadOnlyList<string> _services;
    private readonly VitalsSampler _vitals = new();

    public DockerAdapter(ProjectConfig config)
    {
        _config = config;
        _composeFile = config.ComposeFile ?? "docker-compose.yml";
        _services = config.ComposeServices;
    }

    public void Start() => Compose(["up", "-d", .._services]);

    public void Stop() => Compose(["stop", .._services]);

    public ProcessStatus Status()
    {
        var (stdout, _, _) = Compose("ps", "--format", "json");
        var containers = ParsePsJson(stdout);
        if (_services.Count > 0)
        {
            containers = containers.Where(c => c.Service is not null && _services.Contains(c.Service)).ToList();
        }

        var running = containers.Where(c => c.State == "running").ToList();
        if (running.Count == 0) return new ProcessStatus { Running = false };

        var pid = running[0].Id is { } id ? ContainerPid(id) : null;
        var ports = running
            .SelectMany(c => c.Publishers ?? [])
            .Where(p => p.PublishedPort is not null)
            .Select(p => p.PublishedPort!.Value)
            .Distinct()
            .OrderBy(p => p)
            .ToList();

        (double? cpuPercent, double? memoryMb) vitals = pid is { } p ? _vitals.Sample(p) : (null, null);

        return new ProcessStatus
        {
            Running = true,
            Pid = pid,
            Ports = ports,
            CpuPercent = vitals.cpuPercent,
            MemoryMb = vitals.memoryMb,
        };
    }

    public IReadOnlyList<string> Logs(int tail = 100, string? service = null)
    {
        var args = new List<string> { "logs", "--no-color", "--tail", tail.ToString() };
        if (service is not null) args.Add(service);
        else if (_services.Count > 0) args.AddRange(_services);

        var (stdout, stderr, _) = Compose(args.ToArray());
        return (stdout + stderr)
            .Split('\n')
            .Where(line => !string.IsNullOrWhiteSpace(line))
            .ToList();
    }

    public IReadOnlyList<string> Services()
    {
        if (_services.Count > 0) return _services;
        var (stdout, _, _) = Compose("config", "--services");
        return stdout.Split('\n').Where(l => !string.IsNullOrWhiteSpace(l)).ToList();
    }

    private int? ContainerPid(string containerId)
    {
        var psi = new ProcessStartInfo
        {
            FileName = "docker",
            RedirectStandardOutput = true,
            UseShellExecute = false,
        };
        psi.ArgumentList.Add("inspect");
        psi.ArgumentList.Add("-f");
        psi.ArgumentList.Add("{{.State.Pid}}");
        psi.ArgumentList.Add(containerId);

        using var process = Process.Start(psi)!;
        var output = process.StandardOutput.ReadToEnd().Trim();
        process.WaitForExit();

        return int.TryParse(output, out var pid) && pid != 0 ? pid : null;
    }

    private (string Stdout, string Stderr, int ExitCode) Compose(params string[] args)
    {
        var psi = new ProcessStartInfo
        {
            FileName = "docker",
            WorkingDirectory = _config.Path,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
        };
        psi.ArgumentList.Add("compose");
        psi.ArgumentList.Add("-f");
        psi.ArgumentList.Add(_composeFile);
        foreach (var arg in args) psi.ArgumentList.Add(arg);

        using var process = Process.Start(psi)!;
        var stdout = process.StandardOutput.ReadToEnd();
        var stderr = process.StandardError.ReadToEnd();
        process.WaitForExit();

        return (stdout, stderr, process.ExitCode);
    }

    private static List<ComposeContainer> ParsePsJson(string output)
    {
        output = output.Trim();
        if (output.Length == 0) return [];

        try
        {
            if (output.StartsWith('['))
            {
                return JsonSerializer.Deserialize<List<ComposeContainer>>(output, JsonOptions) ?? [];
            }
            return [JsonSerializer.Deserialize<ComposeContainer>(output, JsonOptions)!];
        }
        catch (JsonException)
        {
            // `docker compose ps --format json` às vezes devolve um array, às vezes NDJSON
            // (um objeto por linha) dependendo da versão instalada — mesma ambiguidade que o
            // adapter Python já tratava.
            return output.Split('\n')
                .Where(l => !string.IsNullOrWhiteSpace(l))
                .Select(l => JsonSerializer.Deserialize<ComposeContainer>(l, JsonOptions)!)
                .ToList();
        }
    }
}

internal sealed class ComposeContainer
{
    [JsonPropertyName("ID")]
    public string? Id { get; set; }
    public string? Service { get; set; }
    public string? State { get; set; }
    public List<ComposePublisher>? Publishers { get; set; }
}

internal sealed class ComposePublisher
{
    public int? PublishedPort { get; set; }
}
